### PR TITLE
Get EmptyHttpResult in RDF via reflection

### PR DESCRIPTION
**Pull request by captainsafia**
Addresses https://github.com/dotnet/aspnetcore/issues/45063

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -92,8 +92,13 @@ public static partial class RequestDelegateFactory
...
     private static readonly MemberExpression StatusCodeExpr = Expression.Property(HttpResponseExpr, typeof(HttpResponse).GetProperty(nameof(HttpResponse.StatusCode))!);
     private static readonly MemberExpression CompletedTaskExpr = Expression.Property(null, (PropertyInfo)GetMemberInfo<Func<Task>>(() => Task.CompletedTask));
-    private static readonly NewExpression EmptyHttpResultValueTaskExpr = Expression.New(typeof(ValueTask<object>).GetConstructor(new[] { typeof(EmptyHttpResult) })!, Expression.Property(null, typeof(EmptyHttpResult), nameof(EmptyHttpResult.Instance)));
-
+    // Due to https://github.com/dotnet/aspnetcore/issues/41330 we cannot reference the EmptyHttpResult type
+    // but users still need to assert on it as in https://github.com/dotnet/aspnetcore/issues/45063
+    // so we temporarily work around this here by using reflection to get the actual type.
+    private static readonly NewExpression EmptyHttpResultValueTaskExpr = Type.GetType("Microsoft.AspNetCore.Http.Results, Microsoft.AspNetCore.Http.Results") is {} resultsType
```
Comment (captainsafia) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:98]: This conditional primarily exists to support scenarios where RDF is instantiated in scenarios where the Results assembly isn't referenced (SignalR tests). This shouldn't impact scenarios where users are running this from the framework.

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 2]
```diff
@@ -2154,32 +2159,34 @@ static async Task ExecuteAwaited(ValueTask task)
 
     private static ValueTask<object?> ExecuteTaskWithEmptyResult(Task task)
     {
+        Debug.Assert(EmptyHttpResultInstance is not null);
```
Comment (brunolins16) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:2162]: Is not possible to it be `null` in the `SignalR` scenario?

Comment (captainsafia) [reply src/Http/Http.Extensions/src/RequestDelegateFactory.cs:2162]: Yep, the problem in the SignalR tests was the resolution that was happening for the assembly in the static properties. I cleaned this up to account for Stephen's feedback and reused the same `EmptyHttpResult.Instance` variable when constructing the property.

**Review by halter73 — APPROVED** [review]
I would still prefer if this threw in release builds if `EmptyHttpResultInstance` is null.

**Comment by captainsafia** [comment]
> I would still prefer if this threw in release builds if `EmptyHttpResultInstance` is null.

Ah, I misunderstood what you're previous feedback was. I thought you meant throw exceptions in all builds (which would impact our local test scenario). I can add an if-def to throw on RELEASE builds.

**Pull request merged**